File: examples/std.cpp
#include <array>
#include <cmath>
#include <complex>
#include <span>
#include <vector>

constexpr auto kBlockSize = 4096;
constexpr auto kChannels  = 2;

int main() {
  //============================================================================
  float carray[kBlockSize]{};
  float carray_carray[kChannels][kBlockSize]{};
  float* carray_ptr[kChannels]{carray_carray[0], carray_carray[1]};
  auto** ptr_ptr = carray_ptr;

  //============================================================================
  std::vector<float> carray_vector[kChannels] = {
      std::vector<float>(kBlockSize), std::vector<float>(kBlockSize)};
  std::array<float, kBlockSize> carray_array[kChannels] = {};
  std::span<float> carray_span[kChannels] = {carray_array[1], carray_vector[1]};
  double carray_d[kBlockSize]             = {};
  std::complex<float> carray_cpx[kBlockSize] = {};

  //==============================================================================
  auto array_array  = std::array<std::array<float, kBlockSize>, kChannels>{};
  auto array_vector = std::array<std::vector<float>, kChannels>{
      std::vector<float>(kBlockSize), std::vector<float>(kBlockSize)};
  auto array_span =
      std::array<std::span<float>, kChannels>{array_vector[0], array_array[0]};

  //==============================================================================
  auto vector_array  = std::vector<std::array<float, kBlockSize>>{kChannels};
  auto vector_vector = std::vector<std::vector<float>>{
      std::vector<float>(kBlockSize), std::vector<float>(kBlockSize)};
  auto vector_span =
      std::vector<std::span<float>>{vector_vector[0], vector_array[0]};

  //==============================================================================
  auto span_array  = std::span{array_array};
  auto span_vector = std::span{vector_vector};
  auto span_span   = std::span{vector_span};

  //============================================================================
  auto array      = std::array<float, kBlockSize>{};
  auto cpx_array  = std::array<std::complex<float>, kBlockSize>{};
  auto vector     = std::vector<float>(kBlockSize);
  auto cpx_vector = std::vector<std::complex<float>>(kBlockSize);
  auto span       = std::span(vector);
  auto cpx_span   = std::span(cpx_vector);
  auto& carray_r  = carray;
  auto& array_r   = array;
  auto* ptr       = carray;
  auto* cpx_ptr   = cpx_array.data();

  // Fill with zeros
  std::fill(carray, carray + kBlockSize, 0.F);
  std::fill(array.begin(), array.end(), 0.F);

  // Fill with std::sin
  auto step  = 2.F * 3.14F / kBlockSize * 16;
  auto phase = 0.F;
  for (auto i = 0; i < kBlockSize; ++i) {
    auto val            = std::sin(phase);
    carray[i]           = val;
    carray_carray[0][i] = val;
    carray_array[0][i]  = val * val;
    carray_vector[0][i] = -val * val;
    carray_vector[1][i] = val / 2;
    carray_span[0][i]   = -val;
    carray_d[i]         = val;
    array[i]            = -val;
    vector[i]           = val;
    carray_cpx[i]       = val;
    phase += step;
    if (i % 8 == 0)
      step *= 1.01F;
  }

  // Apply 0.5 gain
  for (auto i = 0; i < kBlockSize; ++i) {
    carray[i] *= 1.5F;
    array[i] *= 0.5F;
    vector[i] *= 0.5F;
    carray_cpx[i] *= 0.5F;
    carray_vector[0][i] *= 0.5F;
    carray_vector[1][i] *= 0.5F;
  }

  return 0;
}

File: examples/rust/src/bin/std.rs
use core::f32;

const BLOCK_SIZE: usize = 4096;
const CHANNELS: usize = 2;

fn main() {
    //============================================================================
    let _test_f64 = 0.0f64;
    let _test_cpx = num_complex::c32(0.0f32, 0.0f32);
    let mut array = [0.0f32; BLOCK_SIZE];
    let mut cpx_array = [num_complex::c32(0.0f32, 0.0f32); BLOCK_SIZE];
    let mut array_array = [[0.0f32; BLOCK_SIZE]; CHANNELS];

    //============================================================================
    let mut vector = vec![0.0f32; BLOCK_SIZE];
    let mut vector_vector = vec![vec![0.0f32; BLOCK_SIZE]; CHANNELS];
    let mut cpx_vector = vec![num_complex::c32(0.0f32, 0.0f32); BLOCK_SIZE];

    //============================================================================
    let mut _array_vector: [Vec<f32>; CHANNELS] = std::array::from_fn(|_| vec![0.0f32; BLOCK_SIZE]);
    let mut _vector_array = vec![[0.0f32; BLOCK_SIZE]; CHANNELS];

    //============================================================================
    let _slice = &mut array[..];
    let _cpx_slice = &mut cpx_array[..];
    let _immut_slice = &vector[..];
    let array_slice = [&array_array[0][..], &array_array[1][..]];
    #[allow(clippy::useless_vec)]
    let _vector_slice = vec![&array_array[0][..], &array_array[1][..]];
    let _slice_slice = &array_slice[..];
    let _slice_array = &array_array[..];
    let _slice_vector = &vector_vector[..];

    //============================================================================

    // Fill with std::sin
    let step_init = 2.0 * std::f32::consts::PI / BLOCK_SIZE as f32 * 16.0;
    let phase_init = 0.0f32;
    (0..BLOCK_SIZE)
        .scan((phase_init, step_init), |(phase, step), i| {
            let val = phase.sin();
            if i % 8 == 0 {
                *step *= 1.01;
            }
            *phase += *step;
            Some(val)
        })
        .zip(array.iter_mut())
        .zip(array_array[0].iter_mut())
        .zip(vector.iter_mut())
        .zip(vector_vector[1].iter_mut())
        .zip(cpx_array.iter_mut())
        .zip(cpx_vector.iter_mut())
        .for_each(
            |(
                (
                    ((((val, array_elm), array_array_elm), vector_elm), vector_vector_elm),
                    cpx_array_elm,
                ),
                cpx_vector_elm,
            )| {
                *array_elm = val;
                *array_array_elm = val * val;
                *vector_elm = -val;
                *vector_vector_elm = -val * val;
                cpx_array_elm.re = val;
                cpx_array_elm.im = -val * val;
                cpx_vector_elm.re = val * val;
                cpx_vector_elm.im = -val;
            },
        );

    // Apply 0.5 gain
    array[0] = 1.0f32;
    array.iter_mut().for_each(|val| *val *= 0.5f32);
    array_array[0].iter_mut().for_each(|val| *val *= 0.5f32);
    vector.iter_mut().for_each(|val| *val *= 0.5f32);
    vector_vector[1].iter_mut().for_each(|val| *val *= 0.5f32);

    // Test some NaN/inf values
    array[0] = f32::NAN;
    array[BLOCK_SIZE / 2] = f32::INFINITY;
    array[BLOCK_SIZE - 1] = -f32::INFINITY;
    vector[0] = f32::NAN;
    vector[BLOCK_SIZE / 2] = f32::INFINITY;
    vector[BLOCK_SIZE - 1] = -f32::INFINITY;
    array_array[0][0] = f32::NAN;
    array_array[0][BLOCK_SIZE / 2] = f32::INFINITY;
    array_array[0][BLOCK_SIZE - 1] = -f32::INFINITY;
    vector_vector[1][0] = f32::NAN;
    vector_vector[1][BLOCK_SIZE / 2] = f32::INFINITY;
    vector_vector[1][BLOCK_SIZE - 1] = -f32::INFINITY;
}
